### PR TITLE
36 implement plugin wallet endpoints

### LINKED ISSUE #36: Implement plugin wallet endpoints
new endpoints where emplemented here: https://github.com/massalabs/thyra-plugin-wallet/blob/main/api/walletApi-V0.yml

We should update our code to match those endpoints.

This Issue will be blocked until a new release of the plugin wallet (to be synch with @SlnPons for the release and with @Thykof for how to test the code using the main branche of tha repo)

### PR DESCRIPTION
**Pull request by Elli610**
- Problem for importAccount: according to the yml there are no aguments whereas we are supposed to give 2 (pubkey and privkey) -> when we call the method, a popup opens and we enter the secretkey
Should we update the wallet-provider or the thyra-plugin-wallet api ? @gregLibert 

The following methods from wallet-provider need to be added into the thyra-plugin-wallet api:
- buy_rolls
- sell_rolls
- sendTransaction

closes #36

### CODE REVIEW

**Comment by gregLibert** [comment]
Why do you need verifySignature ?

**Comment by gregLibert** [comment]
> * Problem for importAccount: according to the yml there are no aguments whereas we are supposed to give 2 (pubkey and privkey) -> when we call the method, a popup opens and we enter the secretkey
>   Should we update the wallet-provider or the thyra-plugin-wallet api ? @gregLibert

Nope, it simply means that thyra doesn't allow right now to import an account.

File: src/thyra/ThyraProvider.ts [hunk 1]
```diff
@@ -20,13 +22,15 @@ import { IAccountDetails } from '../account';
 /**
  * The Thyra accounts url
  */
-export const THYRA_ACCOUNTS_URL =
-  'https://my.massa/thyra/plugin/massalabs/wallet/rest/wallet';
+export const THYRA_BASE_URL =
+  'https://my.massa/thyra/plugin/massalabs/wallet/rest/wallet/';
+
+export const THYRA_URL = 'http://my.massa';
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:28]: why not using https ?

File: src/thyra/ThyraProvider.ts [hunk 2]
```diff
@@ -20,13 +22,15 @@ import { IAccountDetails } from '../account';
 /**
  * The Thyra accounts url
  */
-export const THYRA_ACCOUNTS_URL =
-  'https://my.massa/thyra/plugin/massalabs/wallet/rest/wallet';
+export const THYRA_BASE_URL =
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:25]: should be THYRA_WALLET_URL

File: src/thyra/ThyraProvider.ts [hunk 3]
```diff
@@ -114,8 +116,27 @@ export class ThyraProvider implements IProvider {
...
+      privateKey,
+    };
+    let thyraAccountsResponse: JsonRpcResponseData<unknown> = null;
+    try {
+      thyraAccountsResponse = await putRequest<unknown>(
+        `${THYRA_BASE_URL}api/accounts`,
+        accountImportRequest,
+      );
+    } catch (ex) {
+      console.error(`Thyra accounts retrieval error`);
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:130]: you should log the error. Here you are hiding it.

File: src/thyra/ThyraProvider.ts [hunk 4]
```diff
@@ -138,7 +159,7 @@ export class ThyraProvider implements IProvider {
     if (allAccounts.isError || allAccounts.error) {
       throw allAccounts.error.message;
     }
-
+    console.log(`allAccounts`, allAccounts);
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:162]: to debug your code ?

File: src/thyra/ThyraProvider.ts [hunk 5]
```diff
@@ -114,8 +116,27 @@ export class ThyraProvider implements IProvider {
...
-    // TODO: once  massalabs/thyra-plugin-wallet#82 is merged, to be updated here
-    throw new Error(`Unimplemented method!`);
+    const accountImportRequest: IAccountImportRequest = {
+      publicKey,
+      privateKey,
+    };
+    let thyraAccountsResponse: JsonRpcResponseData<unknown> = null;
+    try {
+      thyraAccountsResponse = await putRequest<unknown>(
+        `${THYRA_BASE_URL}api/accounts`,
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:126]: Why not using THYRA_ACCOUNTS_URL isntead ?

File: src/thyra/ThyraProvider.ts [hunk 6]
```diff
@@ -148,7 +169,7 @@ export class ThyraProvider implements IProvider {
     let thyraAccountsResponse: JsonRpcResponseData<unknown> = null;
     try {
       thyraAccountsResponse = await deleteRequest<unknown>(
-        `${THYRA_ACCOUNTS_URL}/${accountToDelete.nickname}`,
+        `${THYRA_BASE_URL}/api/accounts/${accountToDelete.nickname}`,
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:172]: Why not using THYRA_ACCOUNTS_URL instead ?

File: src/thyra/ThyraProvider.ts [hunk 7]
```diff
@@ -176,7 +197,19 @@ export class ThyraProvider implements IProvider {
...
   public async getNodesUrls(): Promise<string[]> {
-    throw new Error('Method not implemented.');
+    let nodesResponse: JsonRpcResponseData<unknown> = null;
+    try {
+      nodesResponse = await getRequest<unknown>(`${THYRA_URL}/massa/node`);
+      if (nodesResponse.isError || nodesResponse.error) {
+        throw nodesResponse.error.message;
+      }
+      // transform nodesResponse.result to a json and then get the "url" property
+      const nodes = JSON.parse(JSON.stringify(nodesResponse.result));
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:207]: Why stringifying it to parse it after ?

File: src/thyra/ThyraProvider.ts [hunk 8]
```diff
@@ -185,6 +218,26 @@ export class ThyraProvider implements IProvider {
    * @returns a Promise that resolves to the details of the newly generated account.
    */
   public async generateNewAccount(name: string): Promise<IAccountDetails> {
-    throw new Error('Method not implemented.');
+    let generateNewAccountResponse: JsonRpcResponseData<IThyraWallet> = null;
+    try {
+      console.log(`generateNewAccount:`, THYRA_ACCOUNTS_URL + name);
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:223]: debug ?

File: src/thyra/ThyraProvider.ts [hunk 9]
```diff
@@ -185,6 +218,26 @@ export class ThyraProvider implements IProvider {
    * @returns a Promise that resolves to the details of the newly generated account.
    */
   public async generateNewAccount(name: string): Promise<IAccountDetails> {
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:220]: Considering your remark about the difference between the standard and what the wallet plugin is exposing, I guess that you should keep this function as unimplemented.

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Comment by Elli610** [comment]
> Why do you need verifySignature ?

we don't

File: src/thyra/ThyraProvider.ts (hunk 1 above)
Comment (Elli610) [reply src/thyra/ThyraProvider.ts:28]: If it is 'https', the script cannot connect to thyra:
"""
Thyra nodes retrieval error Axios Error: Error: unable to verify the first certificate
[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Axios Error: Error: unable to verify the first certificate".] {
   code: 'ERR_UNHANDLED_REJECTION'
}
"""
I have to check if it works on another computer and I forgot to set it back to https

File: src/thyra/ThyraProvider.ts [hunk 10]
```diff
@@ -176,7 +196,20 @@ export class ThyraProvider implements IProvider {
    * @returns a Promise that resolves to a list of node urls.
    */
   public async getNodesUrls(): Promise<string[]> {
-    throw new Error('Method not implemented.');
+    let nodesResponse: JsonRpcResponseData<unknown> = null;
+    try {
+      console.log(`${THYRA_URL}/massa/node`);
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:201]: Debug log to be removed.

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: src/thyra/ThyraProvider.ts [hunk 11]
```diff
@@ -114,8 +116,27 @@ export class ThyraProvider implements IProvider {
...
+      privateKey,
+    };
+    let thyraAccountsResponse: JsonRpcResponseData<unknown> = null;
+    try {
+      thyraAccountsResponse = await putRequest<unknown>(
+        THYRA_ACCOUNTS_URL,
+        accountImportRequest,
+      );
+    } catch (ex) {
+      console.log(`Thyra accounts retrieval error`);
```
Comment (gregLibert) [inline src/thyra/ThyraProvider.ts:130]: can you add the error content?

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Review by gregLibert — APPROVED** [review]

**Pull request merged**